build-utils/mount-partitions.sh

```diff
@@ -0,0 +1,100 @@
+#!/bin/sh
+
+# Mounts boot, state, & data partitions from balenaOS.
+# The container must be privileged for this to function correctly.
+
+use_legacy_message="Using legacy host mounts."
+
+# Get the current boot device in case there are duplicate partition labels
+# for `(balena|resin)-(boot|state|data)` found. Because Supervisor is started with
+# `--net=host`, /etc/resolv.conf is bind-mounted by the Engine into the container,
+# so the partition where resolv.conf is located is assumed to be one of the
+# partitions of the current active device. The partition identifer `pN` is removed 
+# from the end so we can compare with the base device.
+current_boot_device=$(grep /etc/resolv.conf /proc/mounts | awk '{print substr($1,1,length($1)-2)}')
+if [ "${current_boot_device}" = "" ]; then
+    echo "WARNING: Could not determine boot device. Please launch Supervisor as a privileged container with host networking."
+fi
+
+# Mounts a device to a path if it's not already mounted.
+# Usage: do_mount DEVICE MOUNT_PATH 
+do_mount() {
+    device=$1
+    path=$2
+
+    # Create the directory if it doesn't exist
+    mkdir -p "${path}"
+
+    # Mount the device if it doesn't exist
+    if [ "$(mountpoint -n "${path}" | awk '{ print $1 }')" != "${device}" ]; then
+        mount "${device}" "${path}"
+    fi
+}
+
+# Find the devices for each balenaOS partition.
+# Usage: setup_then_mount PARTITION MOUNT_PATH
+# PARTITION should be one of boot, state, or data. 
+setup_then_mount() {
+    # If in test environment, pretend we've succeeded at mounting everything to their
+    # new mountpoints. We don't want to actually mount in a containerized test environment 
+    # where the Supervisor is probably not running on a host that has the needed partitions.
+    if [ "${TEST}" = 1 ]; then
+        return 0
+    fi
+
+    partition=$1
+    path=$2
+
+    # Get one or more devices matching label, accounting for legacy partition labels.
+    device=$(blkid | grep -E "(resin|balena)-${partition}" | awk -F':' '{print $1}')
+    
+    count=$(echo "${device}" | grep -c "/dev")
+    if [ "${count}" = 0 ]; then
+        # If no devices with the partition label are found, use legacy mounts.
+        echo "WARNING: No partitions labeled '(resin|balena)-${partition}' found. $use_legacy_message"
+        return 1
+    fi
+
+    if [ "${count}" = 1 ]; then
+        # QUESTION: Should we check whether do_mount succeeded in mounting and inform of failure?
```

Comment:
The multiple device case seems to already includes the single device case, you might be able to remove this and simplify the code

Reply:
Simplified in latest force-push.